Overlay Tunnel Endpoint Manual Setting by CIDR

TrevorTaoARM opened the pull request:
This feature enhancement enables the user to set the desired
outgoing interface and its tunnel endpoint IP(v4) address by
a designated CIDR, which may be different from the default
which currently is just the Kubernetes node IP.

The usage of the this feature is just by adding the following
declaration in the installation yaml file:

apiVersion: v1
kind: ConfigMap
metadata:
  name: cilium-config
  namespace: kube-system
data:
  ...
  tunnel-endpoint-ipv4-cidr: "xxx.xxx.xxx.xxx/yyy"

This feature supports VXLAN overlay tunnel, which would lead the
data traffic and tunnel selection from the default k8s node IP
and related interface to the IP address which matches the setted
IPv4 CIDR and its interface.

If this feature is enabled by setting the desired CIDR,
the MTU and NodePort (physical) device would also be changed to this
matching interface(device) to enable the working of BPF on it.

This feature currently doesn't support IPSEC or node encryption.
This feature is still in its initial state currently which just
support 1 CIDR specified, and the multi-CIDRs case would be
supported if needed.

For further information, please look for:
https://github.com/cilium/cilium/issues/14520

Signed-off-by: trevor tao <trevor.tao@arm.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

Fixes: #14520

```release-note
Add a new feature to enable setting/selecting the overlay tunnel endpoint IPv4 address manually by specifying CIDR
```

joestringer commented on pkg/option/config.go:
```diff
@@ -763,6 +763,9 @@ const (
 	// IPv4NativeRoutingCIDR describes a CIDR in which pod IPs are routable
 	IPv4NativeRoutingCIDR = "native-routing-cidr"
 
+	// TunnelEndpointIPv4CIDR describes a CIDR which the tunnel endpoint IP lies on
+	TunnelEndpointIPv4CIDR = "tunnel-endpoint-ipv4-cidr"
```
I don't think we need to restrict this to tunnel, we could name this like `primary-network-cidr` and use that for the administrator to declare the intent that the network with this CIDR should be the primary network used for in-cluster traffic. For now, we can limit its usage in conjunction with `--tunnel` enabled (ie, set to vxlan/geneve) and otherwise leave the bulk of the PR the same. But then at least if this type of control is desired in future, we can use the same parameter regardless of whether you enable tunnel mode or not.

TrevorTaoARM replied:
Yes, I would consider to provide an option like "primary-network-cidr" as you mentioned in the following work, which would support both the tunnel and direct-routing modes.  Or as you suggested, using that option but limit its uage.

joestringer commented on pkg/cidr/cidr.go:
```diff
@@ -135,3 +135,55 @@ func MustParseCIDR(str string) *CIDR {
+func DetectCIDRByCIDR(cidr *CIDR, version int) (*net.Interface, net.IP, *net.IPNet, error) {
+	netIfaces, err := net.Interfaces()
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("Can't detect the CIDR: Failed to enumerate interfaces")
```
This is currently obscuring the real underlying error, please wrap errors instead:

```suggestion
		return nil, nil, nil, fmt.Errorf("Can't detect the CIDR: Failed to enumerate interfaces: %w", err)
```

This is not the only such case, please inspect all error paths to avoid losing errors.

joestringer commented on pkg/cidr/cidr.go:
```diff
@@ -135,3 +135,55 @@ func MustParseCIDR(str string) *CIDR {
+			if err != nil {
+				continue
+			}
+			if IPVersion(&ip) == version && ip.IsGlobalUnicast() && matchCIDRs(ip, cidrs) {
```
I think that the `IPVersion()` usage here is pushing user input validation too late. AFAICT this PR only supports IPv4, so let's just put a `ParseCIDR(...)` for the user input + `To4()` check really early once, then we don't need this `IPVersion(...)` logic.

joestringer commented on pkg/cidr/cidr.go:
```diff
@@ -135,3 +135,55 @@ func MustParseCIDR(str string) *CIDR {
+	return false
+}
+
+func DetectCIDRByCIDR(cidr *CIDR, version int) (*net.Interface, net.IP, *net.IPNet, error) {
```
The users of this function only check the specific IP and the error, so overall this function seems to be attempting the same process as `firstGlobalAddr()` from `pkg/node`, [see here](https://github.com/cilium/cilium/blob/265d6775c2e0eb4e9d1dfef3ebeaa19a9b976e5f/pkg/node/address_linux.go#L34). Is there anything about the behaviour of that function which is incorrect compared to what this function is trying to achieve? I'd like to fix up the one implementation and reuse it rather than reimplementing similar functionality in multiple places if possible.

TrevorTaoARM replied:
Trevor>> I checked firstGlobalAddr(), but it is for finding a preferredIP in a certain interface but not for matching CIDR. I think the matching/detecing CIDR functions are still natural and should not causing any bothering here.
Anyway, I would check to re-use the existing functions again if possible.

joestringer commented on pkg/datapath/ipcache/listener.go:
```diff
@@ -162,9 +162,13 @@ func (l *BPFListener) OnIPIdentityCacheChange(modType ipcache.CacheModification,
 			nodeIPv4 := node.GetIPv4()
-			if ip4 := newHostIP.To4(); ip4 != nil && !ip4.Equal(nodeIPv4) {
-				copy(value.TunnelEndpoint[:], ip4)
+			// If the tunnel endpoint IPv4 address is manually set, don't change value.TunnelEndpoint
+			if tunnelIP := node.GetTunnelEndpointIPv4(); tunnelIP == nil {
+				if ip4 := newHostIP.To4(); ip4 != nil && !ip4.Equal(nodeIPv4) {
+					copy(value.TunnelEndpoint[:], ip4)
+				}
```
Can you explain what you're trying to achieve here? I'm not sure I understand why tunnel should be treated differently.

TrevorTaoARM replied:
The reason here:
1. Originally, when there is NO tunnel-endpoint-manual-setting, the copy() action is executed only when the nodeIP(here nodeIPv4) is changed; if there's no change, then no copy() here;
2. When there is tunnel-endpoint-manual-setting, the value.TunnelEndpoint here is only setted by the manually selected IP address which matches the desired CIDR which should had been setted previously, so there's no change when the nodeIP is changed but now it's in tunnel-endpoint-manual-setting mode.
3.  I said in the PR description that currently , the tunnel endpoint address is setted from the nodeIP by default(in which there is no manual setting). You could see the nodeIP setting in WaitForNodeInformation() function in pkg/k8s/init.go:
		if nodeIP4 != nil {
			node.SetIPv4(nodeIP4)
		}
4. If there's more concern here, free to discuss further.

joestringer commented on pkg/datapath/linux/config/config.go:
```diff
@@ -399,6 +403,11 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 		}
 	}
 
+	// Manually defined a tunnel endpoint IPv4 CIDR
+	if tunnelEndpointIPv4CIDR := option.Config.GetTunnelEndpointIPv4CIDR(); tunnelEndpointIPv4CIDR != nil {
+		cDefinesMap["TUNNEL_ENDPOINT_IPV4_CIDR"] = tunnelEndpointIPv4CIDR.String()
+	}
```
I don't see any BPF code corresponding to this macro declaration. Is there another piece that we're not seeing or is this only for informational purposes?

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1307,6 +1311,18 @@ func initEnv(cmd *cobra.Command) {
+		if _, intfIP, _, err := cidr.DetectCIDRByCIDR(tunnelEndpointIPv4CIDR, 4); err != nil {
+			log.WithField(logfields.CIDR, tunnelEndpointIPv4CIDR.IPNet.String()).Fatal("Can't detect the cidr for an actual interface IP")
+		} else {
+			log.Infof("Now set the node tunnel address: %s", intfIP.String())
```
Two comments here:
* Please avoid Foof() variants on the log functions, instead when possible use `log.WithFields(...).Info(...)`.
* I don't think that this log is necessary at Info level as the user configuration is already printed during startup. Maybe this is good for debugging but in that case please switch to `log.Debug(...)`.

joestringer commented on pkg/option/config.go:
```diff
@@ -3058,6 +3082,19 @@ func (c *DaemonConfig) checkIPv4NativeRoutingCIDR() error {
+// check the tunnel endpoint IPv4 cidr can match an actual interface IP
+func (c *DaemonConfig) checkIPv4TunnelEndpointCIDR() error {
+	tunnelEndpointIPv4CIDR := c.GetTunnelEndpointIPv4CIDR()
+	log.WithField(logfields.CIDR, tunnelEndpointIPv4CIDR).Info("Check Tunnel Endpoint IPv4 CIDR")
```
:+1: better use of `log.WithField(..)` / `log.WithFields(...)` here but again this looks more like debug logging rather than Info that all users may need.

TrevorTaoARM replied:
Thanks!

joestringer commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -207,6 +207,15 @@ func (n *NodeDiscovery) StartDiscovery(nodeName string) {
 
+	// Add the tunnel endpoint address here
+	log.Debug("Adding local node tunnel endpoint address")
+	if node.GetTunnelEndpointIPv4() != nil {
```
Between K8sExternalIPv4, IPv4, IPv4Router, TunnelIPv4 I'm getting confused by which IP addresses will be used in which circumstance. Do we really need to keep track of as many as four different IP addresses here?

joestringer replied:
/cc @AnishShah any thoughts here? Do you see semantic overlap with the existing address fields?

joestringer replied:
My gut feeling on first reading this is that the option you're trying to specify here should be able to configure the `NodeInternalIP` and let those codepaths handle the traffic rather than adding a new dedicated tunnel option which is sometimes specified, sometimes not.

TrevorTaoARM replied:
Thanks for questions here. I think each address type here is an needed one. Maybe we can extend the use of TunnelIPv4 here to separate the data path and control path in the future. But it actually provide us something now.
Other comments:
https://github.com/cilium/cilium/pull/15072#issuecomment-796542640

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -368,7 +368,11 @@ func (m *Manager) NodeUpdated(n nodeTypes.Node) {
 			option.Config.EnableHostFirewall || option.Config.JoinCluster {
-			tunnelIP = nodeIP
+			if tunnelIP = n.GetNodeTunnelEndpointIPv4(); tunnelIP == nil {
+				log.Debugf("No tunnelIP manually defined, just use the nodeIP")
```
nit: `Debugf()` uses the format variation, which implies that there are parameters coming. However, this is only a string. This could be equally done with `log.Debug(...)`.

TrevorTaoARM replied:
Removed the log function here.

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -368,7 +368,11 @@ func (m *Manager) NodeUpdated(n nodeTypes.Node) {
+				log.Debugf("No tunnelIP manually defined, just use the nodeIP")
+				tunnelIP = nodeIP
+			}
+			log.Debugf("tunnelIP %s", tunnelIP.String())
```
nit: Even if debug is only seen by developers, if you intend to merge it into the codebase then please format the message in a way that a developer could understand it; printing random values in the debug logs makes the message difficult to interpret.

TrevorTaoARM replied:
Removed the debug log here.

joestringer commented on pkg/datapath/linux/node.go:
```diff
@@ -877,9 +880,29 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+			oldTunnelIP4 = oldIP4
+			newTunnelIP4 = newIP4
+		}
+		log.WithField("tunnelIP", newTunnelIP4.String()).Info("Set tunnel endpoint IPv4 address")
```
:thinking:

TrevorTaoARM replied:
Removed the log functions here.

joestringer reviewed (changes requested):
Thanks for the PR!

My main question for this PR is how this new IP fits into the list of all of the other IPs that we're tracking internally in Cilium. Do we really need to track four different IP addresses independently? Now that we have a dedicated K8sExternalIP that we track in the nodes, I feel like we might be able to manually repurpose the InternalIP for cluster-internal traffic like this. If this new option could override the InternalIP, then I don't think we would need to add another tunnel-specific option that is only sometimes used. Did you consider that option, is it a reasonable approach? See below for the thread, we can discuss there.

Ideally, I would prefer if we don't need to restrict this to tunnels, we could consider calling the option `--primary-network-cidr` like I suggested in #14339 and use this as the primary IP that we use to derive settings like this. For now we can restrict to IPv4 + tunnel mode configured, but in future we could relax it if administrators wanted more control over which networks are being used for node<->node connectivity.

I noticed some places where it seemed like there is similar code already elsewhere in the codebase where we could share the code better. 

Other than that, mostly minor nits that should be easy to resolve.

joestringer commented on pkg/cidr/cidr.go:
```diff
@@ -135,3 +135,55 @@ func MustParseCIDR(str string) *CIDR {
+}
+
+func DetectCIDRByCIDR(cidr *CIDR, version int) (*net.Interface, net.IP, *net.IPNet, error) {
+	netIfaces, err := net.Interfaces()
```
FYI if we do not remove this code while addressing the other feedback on this PR, then we will want to consider replacing `net.Interfaces()` here with `netlink.LinkList()` for the same reason as https://github.com/cilium/cilium/pull/15260.

TrevorTaoARM replied:
Updated accordingly. Thanks

jibi committed:
iptables: relax no CT rules to match all pod traffic

The --install-no-conntrack-iptables-rules option was originally intended
to be enabled by default. Because of this, it was targeting only
pod-to-pod traffic to avoid any unpredictable interaction with traffic
to/from the hostns.

Since the original decision of making this option enabled by default has
been revised (now the user has to explicitly enable it) we can relax the
rules to match also pod <-> hostns traffic.

With this commit, the set of Iptables rules installed on all nodes
when the option is enabled changes from:

    iptables -t raw -I CILIUM_PRE_raw    -s $POD_CIDR -d $POD_CIDR -j NOTRACK
    iptables -t raw -I CILIUM_OUTPUT_raw -s $POD_CIDR -d $POD_CIDR -j NOTRACK

(i.e. all traffic which is at the same time originating from and
destined to a pod)

to:

    iptables -t raw -I CILIUM_PRE_raw    -s $POD_CIDR -j NOTRACK
    iptables -t raw -I CILIUM_PRE_raw    -d $POD_CIDR -j NOTRACK
    iptables -t raw -I CILIUM_OUTPUT_raw -s $POD_CIDR -j NOTRACK
    iptables -t raw -I CILIUM_OUTPUT_raw -d $POD_CIDR -j NOTRACK

(i.e. all traffic which is either originating from a pod or destined to
a pod).

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>
Signed-off-by: TrevorTaoARM <trevor.tao@arm.com>
